### PR TITLE
Create onlyoffice extension

### PR DESCRIPTION
**Pull request by LukasHirt**
Requires https://github.com/owncloud/phoenix/pull/4324 to work in ocis-web. Testing requires running oc10 server with the onlyoffice app installed and enabled + running onlyoffice docs server.

Some summary of my brain trips when creating this:

Originally I was considering creating a component which would contain an iframe loading the editor so we are not really leaving ocis-web. This had two problems though - we would need to set a new header to allow this and we would anyway see the topbar of oc10. That's why I decided to go the way of opening the editor in a new tab.

Regarding tests - I don't really see any value in tests here. We do not have any views, any big logic... If the handler is working should be tested in ocis-web on a more general level. Testing if `window.open` seems like a needless increase of CI time.

### CODE REVIEW

**Comment by LukasHirt** [comment]
Just realising now. I've actually added the extension as default into the config.json in ocis-phoenix which doesn't make sense because it won't work yet in pure ocis setup. Need to think about how do we want to deliver this in bridge setup.

File: onlyoffice/.drone.star [hunk 1]
```diff
@@ -0,0 +1,698 @@
+def main(ctx):
```
Comment (kulmann) [inline onlyoffice/.drone.star:1]: This file is not needed. The extension has to be added to the top level `.drone.star` instead.

File: onlyoffice/.github/config.yml [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# Configuration for update-docs - https://github.com/behaviorbot/update-docs
```
Comment (kulmann) [inline onlyoffice/.github/config.yml:1]: file is not needed

File: onlyoffice/.editorconfig [hunk 3]
```diff
@@ -0,0 +1,35 @@
+# http://editorconfig.org
```
Comment (kulmann) [inline onlyoffice/.editorconfig:1]: file is not needed

File: onlyoffice/.github/settings.yml [hunk 4]
```diff
@@ -0,0 +1,98 @@
+---
```
Comment (kulmann) [inline onlyoffice/.github/settings.yml:1]: file is not needed.

File: onlyoffice/.gitignore [hunk 5]
```diff
@@ -0,0 +1,8 @@
+coverage.out
```
Comment (kulmann) [inline onlyoffice/.gitignore:1]: file is not needed. Have a look at the top level `.gitignore` if something needs to be added.

File: onlyoffice/LICENSE [hunk 6]
```diff
@@ -0,0 +1,202 @@
+
```
Comment (kulmann) [inline onlyoffice/LICENSE:1]: we have the license file in all other extension folders as well, but IMO it should only be needed on the top level of the repository.

File: onlyoffice/Makefile [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+	cd $(HUGO); hugo
+
+.PHONY: docs
+docs: docs-copy docs-build
+
+.PHONY: watch
+watch:
+	go run github.com/cespare/reflex -c reflex.conf
+
+# $(GOPATH)/bin/protoc-gen-go:
```
Comment (kulmann) [inline onlyoffice/Makefile:158]: Remove the protoc and protobuf parts. Not needed to have it commented. We have enough other references to copy it over later.

File: onlyoffice/README.md [hunk 8]
```diff
@@ -0,0 +1,40 @@
+# ownCloud Infinite Scale: ONLYOFFICE
```
Comment (kulmann) [inline onlyoffice/README.md:1]: file is not needed

File: onlyoffice/changelog/CHANGELOG.tmpl [hunk 9]
```diff
@@ -0,0 +1,53 @@
+{{ $allVersions := . }}
```
Comment (kulmann) [inline onlyoffice/changelog/CHANGELOG.tmpl:1]: file is not needed

File: onlyoffice/changelog/README.md [hunk 10]
```diff
@@ -0,0 +1,6 @@
+# Changelog
```
Comment (kulmann) [inline onlyoffice/changelog/README.md:1]: file is not needed

File: onlyoffice/changelog/TEMPLATE [hunk 11]
```diff
@@ -0,0 +1,11 @@
+Bugfix: Fix behavior for foobar (in present tense)
```
Comment (kulmann) [inline onlyoffice/changelog/TEMPLATE:1]: file is not needed

File: onlyoffice/cmd/ocis-onlyoffice/main.go [hunk 12]
```diff
@@ -0,0 +1,13 @@
+package main
```
Comment (kulmann) [inline onlyoffice/cmd/ocis-onlyoffice/main.go:1]: rename folder `onlyoffice/cmd/ocis-onlyoffice` to `onlyoffice/cmd/onlyoffice`

File: onlyoffice/docs/_index.md [hunk 13]
```diff
@@ -0,0 +1,10 @@
+---
```
Comment (kulmann) [inline onlyoffice/docs/_index.md:1]: docs should live in the top level `docs` folder

File: onlyoffice/docs/building.md [hunk 14]
```diff
@@ -0,0 +1,28 @@
+---
```
Comment (kulmann) [inline onlyoffice/docs/building.md:1]: docs should live in the top level docs folder

File: onlyoffice/docs/getting-started.md [hunk 15]
```diff
@@ -0,0 +1,272 @@
+---
```
Comment (kulmann) [inline onlyoffice/docs/getting-started.md:1]: docs should live in the top level docs folder

File: onlyoffice/docs/building.md [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+All required tool besides Go itself and make are bundled or getting automatically installed within the `GOPATH`. All commands to build this project are part of our `Makefile`.
+
+## Backend
+
+{{< highlight txt >}}
+make generate
+make build
+{{< / highlight >}}
+
+Finally you should have the binary within the `bin/` folder now, give it a try with `./bin/ocis-onlyoffice -h` to see all available options.
```
Comment (kulmann) [inline onlyoffice/docs/building.md:28]: `./bin/onlyoffice` instead of `./bin/ocis-onlyoffice`

File: onlyoffice/docs/license.md [hunk 17]
```diff
@@ -0,0 +1,10 @@
+---
```
Comment (kulmann) [inline onlyoffice/docs/license.md:1]: docs should live in the top level docs folder

File: onlyoffice/package.json [hunk 18]
```diff
@@ -0,0 +1,78 @@
+{
+  "private": true,
+  "name": "ocis-onlyoffice",
```
Comment (kulmann) [inline onlyoffice/package.json:3]: rename to `onlyoffice`

File: onlyoffice/pkg/command/root.go [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+)
+
+// Execute is the entry point for the ocis-onlyoffice command.
+func Execute() error {
+	cfg := config.New()
+
+	app := &cli.App{
+		Name:     "ocis-onlyoffice",
+		Version:  version.String,
+		Usage:    "Onlyoffice ocis extension",
```
Comment (kulmann) [inline onlyoffice/pkg/command/root.go:22]: Change to `OnlyOffice ocis extension` (branding)

File: onlyoffice/pkg/command/root.go [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+	"github.com/owncloud/ocis/onlyoffice/pkg/version"
+	"github.com/spf13/viper"
+)
+
+// Execute is the entry point for the ocis-onlyoffice command.
+func Execute() error {
+	cfg := config.New()
+
+	app := &cli.App{
+		Name:     "ocis-onlyoffice",
```
Comment (kulmann) [inline onlyoffice/pkg/command/root.go:20]: Change to `onlyoffice`

File: onlyoffice/pkg/service/v0/service.go [hunk 21]
```diff
@@ -0,0 +1,52 @@
...
+	mux    *chi.Mux
+}
+
+// ServeHTTP implements the Service interface.
+func (g Onlyoffice) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	g.mux.ServeHTTP(w, r)
+}
+
+// Dummy implements the Service interface.
+func (g Onlyoffice) Dummy(w http.ResponseWriter, r *http.Request) {
```
Comment (kulmann) [inline onlyoffice/pkg/service/v0/service.go:47]: Not needed

File: onlyoffice/pkg/service/v0/tracing.go [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+	next Service
+}
+
+// ServeHTTP implements the Service interface.
+func (t tracing) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	t.next.ServeHTTP(w, r)
+}
+
+// Dummy implements the Service interface.
+func (t tracing) Dummy(w http.ResponseWriter, r *http.Request) {
```
Comment (kulmann) [inline onlyoffice/pkg/service/v0/tracing.go:24]: Not needed

File: onlyoffice/ui/dev/public/index.html [hunk 23]
```diff
@@ -0,0 +1,17 @@
+<!DOCTYPE html>
+<html lang="en">
+  <head>
+    <meta charset="utf-8">
+    <meta http-equiv="X-UA-Compatible" content="IE=edge">
+    <meta name="viewport" content="width=device-width,initial-scale=1.0">
+    <title>Extension Demo</title>
```
Comment (kulmann) [inline onlyoffice/ui/dev/public/index.html:7]: Replace `Extension Demo` with `OnlyOffice` in this entire file, if the file is needed.

File: onlyoffice/pkg/service/v0/instrument.go [hunk 24]
```diff
@@ -0,0 +1,30 @@
...
+	metrics *metrics.Metrics
+}
+
+// ServeHTTP implements the Service interface.
+func (i instrument) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	i.next.ServeHTTP(w, r)
+}
+
+// Dummy implements the Service interface.
+func (i instrument) Dummy(w http.ResponseWriter, r *http.Request) {
```
Comment (kulmann) [inline onlyoffice/pkg/service/v0/instrument.go:28]: not needed

File: onlyoffice/pkg/service/v0/logging.go [hunk 25]
```diff
@@ -0,0 +1,30 @@
...
+	logger log.Logger
+}
+
+// ServeHTTP implements the Service interface.
+func (l logging) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	l.next.ServeHTTP(w, r)
+}
+
+// Dummy implements the Service interface.
+func (l logging) Dummy(w http.ResponseWriter, r *http.Request) {
```
Comment (kulmann) [inline onlyoffice/pkg/service/v0/logging.go:28]: Not needed

File: onlyoffice/pkg/service/v0/service.go [hunk 26]
```diff
@@ -0,0 +1,52 @@
...
+	"net/http"
+
+	"github.com/go-chi/chi"
+	"github.com/owncloud/ocis/onlyoffice/pkg/config"
+)
+
+// Service defines the extension handlers.
+type Service interface {
+	ServeHTTP(http.ResponseWriter, *http.Request)
+	Dummy(http.ResponseWriter, *http.Request)
```
Comment (kulmann) [inline onlyoffice/pkg/service/v0/service.go:13]: Not needed

File: onlyoffice/pkg/service/v0/service.go [hunk 27]
```diff
@@ -0,0 +1,52 @@
...
+
+	m := chi.NewMux()
+	m.Use(options.Middleware...)
+
+	svc := Onlyoffice{
+		config: options.Config,
+		mux:    m,
+	}
+
+	m.Route(options.Config.HTTP.Root, func(r chi.Router) {
```
Comment (kulmann) [inline onlyoffice/pkg/service/v0/service.go:28]: Not needed

**Review by kulmann — CHANGES_REQUESTED** [review]
I discovered some files that are not needed, some files that need to be one folder level higher, some occurrences of `ocis-onlyoffice` and some code that can be deleted.

Also, please contact @C0rby on how to implement the version command here and how to add it to the boilr template, so that we have it in every new service.

Good to see this extension! Really nice progress.

**Comment by LukasHirt** [comment]
@kulmann I've addressed all your comments. Regarding the versions command, I would like to tackle that in separate PR to finally get this one merged.

**Review by kulmann — APPROVED** [review]
Fix the staticcheck, then it's good to go 🚀

**Pull request merged**